### DISCUSSION TITLE
What does "because this is an alternate version of another formula." mean?

### DISCUSSION THREAD

**Discussion by albas99** [Homebrew/discussions#4318] [Everyday usage] [answered]
### Output of `brew config`

```shell
: 4.0.6
ORIGIN: https://github.com/Homebrew/brew
HEAD: 0c1031f2fd71016f0f5df699caaed384487bfb16
Last commit: 24 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 2a72f5600dfb622613a6fb43403475e9d42f0a02
Core tap last commit: 3 weeks ago
Core tap branch: master
Core tap JSON: 11 Mar 09:10 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: quad-core 64-bit kabylake
Clang: 14.0.0 build 1400
Git: 2.39.2 => /usr/local/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
```

### Output of `brew doctor`

```shell
note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Broken symlinks were found. Remove them with `brew cleanup`:
  /usr/local/bin/apm
  /usr/local/bin/fleet
  /usr/local/lib/docker/cli-plugins
  /usr/local/share/man/man5/package-locks.5
  /usr/local/share/man/man5/shrinkwrap-json.5
```

### Description of issue

I'm installing `postgresql@15` and I got the message `postgresql@15 is keg-only, which means it was not symlinked into /usr/local, because this is an alternate version of another formula.`. I know similar discussions/issues about this message but none of them explained specifically what the phrase "because this is an alternate version of another formula." means. Does it mean this is not the ideal version for my system(Hence the reason why it's not being linked) or something else? The first command I used was just `brew install postgresql` and it installed `postgresql@14` with no issues but I decided to uninstall and install 15 since it's the latest version which gave me the message above.

**Comment by gromgit** [comment, marked as answer]
It simply means that `postgresql@15` is not the primary version of PostgreSQL on Homebrew (`postgresql@14` is), and while you can install multiple PostgreSQL versions side-by-side, There Can Be Only One[^1] version (the primary) that's linked into the main Homebrew directories, where it can be accessed without qualifying paths (i.e. `psql` instead of `${HOMEBREW_PREFIX}/opt/postgresql@14/bin/psql`).

Normally, the primary version is easily identified as the one without an `@` suffix, and it's a proper formula unto itself, upgraded as and when new versions are released by the upstream provider, and tested by Homebrew's infrastructure. For instance, `go` can be v1.19 one day, and v1.20 the next.

This used to be true for `postgresql` too, but automatically upgrading major PostgreSQL versions turned out to cause so many problems that it's now a case of "`postgresql` Considered Harmful",[^2] hence it's now just an _alias_ for `postgresql@14`, and you get this message:
```sh
% brew info postgresql
Warning: Use postgresql@14 instead of deprecated postgresql
...
```
which suggests that the alias will disappear at some point.

[^1]: https://highlander.fandom.com/wiki/There_can_be_only_one
[^2]: http://catb.org/jargon/html/C/considered-harmful.html

**Reply by albas99** [reply]
Got it. I did get that message when I installed `postgresql@14` but using just `brew install postgresql`. I ended up just using `brew link postgresql@15` to just link and that worked out fine. Guess the homebrew team haven't gotten to upgrading the default version to the latest(15).

Thanks for the explanation though.

**Reply by gromgit** [reply]
I can't speak for Homebrew, but like I said, I wouldn't be surprised if default`postgresql` disappears at some point. Users will then have to `brew link postgresql@XYZ` themselves, or add `$(brew --prefix postgresql@XYZ)/bin` to their `PATH`. (The latter has the advantage of allowing different environments on the same machine with different version dependencies.)
